---
title: MacArthur Lab
layout: default
---  
<!-- Page with full research description for the MacArthur Lab -->
<div class="section">
    <h1>MacArthur Lab of MGH and the Broad Institute of MIT and Harvard</h1>  

    <p class="leadText">The <a href="http://www.macarthurlab.org/home/">MacArthur Lab</a> focuses on extracting useful information from human genome sequences. Some of the main focuses of the lab are accurate identification of sequence variants from massive data-sets, identifying LoF variants, using transcriptome sequencing approaches to characterize the impact of DNA sequence variants on human gene function, and using genomic approaches to discover disease-causing mutations in severe disease patients.</p>

    <a id="project2" class="anchor"></a>
    <div class="project">
        <h4 class="projectTitle"><div class="expandIcon">-</div>Gene Models and Annotation Tools</h4>
        <div class="projectBody">
			<p>I will soon lead a project comparing the consequences of using different gene models and tools for annotation of variants from whole exome and whole genome sequencing projects.</p>
			<p>The scope of this work will initially include GENCODE and RefSeq gene models, Variant Effect Predictor, ANNOVAR, and SnpEff annotation tools, and known mutations from HGMD.</p>
			<p>I will also be investigating how to best summarize and consolidate information and annotation from multiple transcripts in GENCODE.</p>
        </div>
    </div>

    <a id="leiden-cleanup" class="anchor"></a>
    <div class="project">
        <h4 class="projectTitle"><div class="expandIcon">-</div>Leiden Muscular Dystrophy Variant Database</h4>
        <div class="projectBody">
            <p>I developed Python scripts to automate extraction and remapping of variants from any Leiden Open Variation Database installation. These databases contain potentially invaluable information for human disease research, but are fraught with inconsistencies and in unusable formats for large-scale data analytics.</p>
			<p>I remapped variants to genomic coordinates, performed extensive validation, and augmented existing information with data from HGMD and allele frequencies from various ethnic populations.</p>
			<p>In the next steps of this project, our group will be working closely with our unit leader Professor Mark Daly to provide data for his work in human disease research.</p>
        </div>
    </div>
</div>






